### ISSUE TITLE
Nice work! Reproducing original results?

### ISSUE THREAD

**Issue by nicklashansen** [ShaneFlandermeyer/tdmpc2-jax#1]
Hi Shane,

Great work on this. I'll definitely have a closer look soon. Do you have any preliminary benchmark results that compare the jax and pytorch versions? Both in terms of env steps and wall-time. Super curious to see how they compare!

**Comment by ShaneFlandermeyer** [comment]
Benchmarks and improved logging are next up on my to-do list. Unfortunately, I have a conference deadline coming up and will have to grind on that for a couple weeks instead.

For now, I'll be a bit more qualitative: I've tested and verified that both implementations perform similarly in the HalfCheetah, Walker, and Humanoid mujoco environments. In these envs, the Jax agent's throughput is 80-90 FPS, while the Torch implementation sits at 15-20

**Comment by nicklashansen** [comment]
Awesome, thanks for the update and best of luck with your deadline!

**Comment by edwhu** [comment]
@ShaneFlandermeyer did you ever get a chance to reproduce more results? Happy to help out here.

**Comment by ShaneFlandermeyer** [comment]
@edwhu Unfortunately I have not. These last couple months have been really busy...

I would **greatly** appreciate help with this if you're up for it! I recommend doing the comparison with the develop branch, as there are some small changes I haven't merged back to main. Let me know if there's anything I can help with!

**Comment by edwhu** [comment]
I'll take a crack at it.

**Comment by edwhu** [comment]
Would be nice to keep a globalized list of feature requests somewhere - it seems like we would like:
- tensorboard logging
- checkpointing
- task embeddings

**Comment by ShaneFlandermeyer** [comment]
Yeah, that just about covers my ideas for a development road map. Logging support and checkpointing are the best places to start imo. I'll add a "contributing" section to the readme with things that can be improved!

**Comment by nicklashansen** [comment]
I don't think I can contribute much code-wise (I don't speak jax) but happy to help out with more conceptual things (if any)!

**Comment by edwhu** [comment]
Are there any good prototypical or debugging tasks you can suggest for comparing performance between the implementations? 

A task that's too easy will let buggy implementations go through. 

Really hard tasks that TDMPC2 excels at are good, but I would like one that's still pretty fast to train on.

**Comment by nicklashansen** [comment]
Yea! I find DMControl tasks to be the most reliable for benchmarking (fairly low variance between seeds), I'd recommend using a handful of tasks from different embodiments so that you don't over-optimize for any single task. Something like `(cheetah-run, walker-run, finger-turn-hard, humanoid-stand)` might be a good task set for experimentation?

On a different note: I'm curious to know exactly what makes the jax implementation of tdmpc2 so much faster? I would be interested in using any insights here to optimize the pytorch version more.

**Comment by edwhu** [comment]
My guess is the JIT compilation of the MPPI planning procedure could make a big difference. It would be interesting to profile both codebases to figure out the actual bottlenecks. 

This JAX implementation is missing some details, documenting here. 
- no grad norm clipping. @nicklashansen  is this important? would also be good to log grad norms for debugging. 
- some heuristics missing like: planning iterations based on action space, adaptive discount factor based on episode length

**Comment by ShaneFlandermeyer** [comment]
> My guess is the JIT compilation of the MPPI planning procedure could make a big difference. It would be interesting to profile both codebases to figure out the actual bottlenecks. 

Yeah, there are just lots of for loops for Jax to take advantage of in both training and planning. I imagine you could get a big performance boost in the main repo by taking advantage of more torch.compile tools
> 
> - some heuristics missing like: planning iterations based on action space, adaptive discount factor based on episode length

I opted to leave these out and let the user define them if they want. Happy to add them in explicitly though!

**Comment by edwhu** [comment]
> Benchmarks and improved logging are next up on my to-do list. Unfortunately, I have a conference deadline coming up and will have to grind on that for a couple weeks instead.
> 
> For now, I'll be a bit more qualitative: I've tested and verified that both implementations perform similarly in the HalfCheetah, Walker, and Humanoid mujoco environments. In these envs, the Jax agent's throughput is 80-90 FPS, while the Torch implementation sits at 15-20.

what GPU are you using? I'm geting around 50 FPS for HalfCheetahV4 on an Nvidia 3090.

**Comment by ShaneFlandermeyer** [comment]
Are you using the 4 envs/UTD=0.25 setting in the config? If so, the number of iterations per second reported by the progress bar is **not** the actual FPS, as each iteration corresponds to num_env environment steps. The bar output is only the FPS for 1 env (50 iter/s is around 200 FPS for 4 envs).

The reported numbers were from a laptop 4090, which I assume is similar to your 3090. If that's the true FPS, I'll dig into it tomorrow and see if I can identify possible performance regressions.

**Comment by edwhu** [comment]
I think you're right. I was wondering why the tqdm was lagging behind the actual global steps.

**Comment by edwhu** [comment]
A **very** preliminary comparison between the jax and original TDMPC2 codebase on DMC Humanoid Stand for 1M steps. Curves look good, but this still isn't completely apples-to-apples. 

<img width="700" alt="Screenshot 2024-05-31 at 4 55 29 PM" src="https://github.com/ShaneFlandermeyer/tdmpc2-jax/assets/13341926/f3e45688-3f43-4b18-a922-6247d5f61483">

One implementation difference is that the Jax implementation uses an async vector env with 4 independent copies, and updates with an UTD of 0.5
I think TDMPC2 has an UTD of 1.0, and not sure if it's running 4 copies of the environment. 

But overall, the trend looks roughly similar. I'm interested in the speed up - around 2hrs for 1M steps, so 8hrs total to hit the 4M steps in the original paper. Although this may change if we increase the UTD from 0.5 to 1.0
@nicklashansen do you have an estimate of the walltime of running TDMPC2 humanoid standup for 1M steps?

**Comment by ShaneFlandermeyer** [comment]
Correct me if I'm wrong @nicklashansen, but the original paper results are 1 env, UTD=1. I'll run 1M steps of Humanoid Stand with those settings from your most recent PR @edwhu. I imagine that'll give curves more similar to what's expected. Progress bar estimates 2.5 hours for 1M steps in 1 env.

**Comment by nicklashansen** [comment]
Yes, the original paper uses 1 env and UTD=1. I believe that the PyTorch implementation is at roughly 1M policy steps (2M sim steps) per day with that setup. Here's some wall-time results comparing 1 env vs. 4 envs: https://github.com/nicklashansen/tdmpc2/issues/18#issuecomment-1939759448

**Comment by ShaneFlandermeyer** [comment]
Here's the 1 env UTD = 1 case. The final reward looks reasonable, but there is some instability in the intermediate steps. Anyone with more DMC experience know if this looks reasonable?

![humanoid-walk_single_env](https://github.com/ShaneFlandermeyer/tdmpc2-jax/assets/13738359/5a8b7ba1-f109-4edc-a5d3-bce4d171d089)

**Comment by edwhu** [comment]
I think this looks reasonable for a single run, although @nicklashansen should confirm. We could also let it run for longer since instability during early training is very normal.

I see there are some [raw TDMPC2 curves here,](https://github.com/nicklashansen/tdmpc2/blob/main/results/humanoid-stand.csv
) could you plot these on top of what you have now?

**Comment by ShaneFlandermeyer** [comment]
Overlayed figure:

![humanoid-walk_single_env_compare](https://github.com/ShaneFlandermeyer/tdmpc2-jax/assets/13738359/ff80e781-35e1-41a2-8afc-f88d75e246e8)

**Comment by edwhu** [comment]
That looks pretty good to me, barring the potential difference in logging between codebases for action repeat.

**Comment by nicklashansen** [comment]
The raw official numbers that we provide are evals, i.e., mean of multiple episodes at fixed intervals. If you're plotting the training episode returns then I would expect to see some variance / performance drops between individual episodes.

**Comment by ShaneFlandermeyer** [comment]
@nicklashansen Did you run into any mujoco physics errors in your DMControl experiments? We've been able to match the PyTorch implementation's performance in all the envs we've tried except for walker-run, which gets a high reward (>850) before throwing a mujoco error.

See #5 for ongoing discussion on the topic.

**Comment by ShaneFlandermeyer** [comment]
Here are the final results for several DMControl envs. The Jax implementation does very well!

![walker-run](https://github.com/ShaneFlandermeyer/tdmpc2-jax/assets/13738359/c307f3bd-1894-434f-a284-a8abd8d6bdc6)
![humanoid-stand](https://github.com/ShaneFlandermeyer/tdmpc2-jax/assets/13738359/f4accdec-dbed-4180-a89e-6583060a77b4)
![finger-turn-hard](https://github.com/ShaneFlandermeyer/tdmpc2-jax/assets/13738359/eb50d2ea-e4c1-491d-ae43-04c21ace088b)
![cheetah-run](https://github.com/ShaneFlandermeyer/tdmpc2-jax/assets/13738359/57372259-b8b7-4360-8503-c26110bf1cce)

**Comment by nicklashansen** [comment]
@ShaneFlandermeyer @edwhu This looks great! Somewhat surprised to see that the performance now exceeds the pytorch version, any idea what has changed?

**Comment by edwhu** [comment]
Might be the REDQ update for the q ensemble.

**Comment by ShaneFlandermeyer** [comment]
@nicklashansen Yep. We tried to match things as closely as possible to the original repo except for in the value estimation and actor loss, where the mean is taken over the full critic ensemble. We have to compute those values anyways due to vmap'ing, so it doesn't add much computation at all.

**Comment by nicklashansen** [comment]
Interesting, I will have to look into this. A few prior works (e.g. REDQ) show that subsampling the Q-functions generally performs better than taking the mean over the full ensemble, but perhaps this is somewhat setup-specific.
